module/zfs/dnode.c

```diff
@@ -1559,6 +1559,7 @@ dnode_hold_impl(objset_t *os, uint64_t object, int flag, int slots,
 	dnode_slots_rele(dnc, idx, slots);
 
 	DNODE_VERIFY(dn);
+	ASSERT3P(dnp, !=, NULL);
```

Comment:
`dnp` is also used on line 1329, so should there be an `IMPLY(!(flag & DNODE_DRY_RUN), (tag != NULL) && (dnp != NULL));` at the top of the function instead?

Reply:
That wouldn't hurt, but I don't think it's really needed in this case.  If you feel strongly about it I can it, but I don't think this has resulted in any real confusion (aside from the cppcheck warning).